pkg/ingest/conn.go

```diff
@@ -55,76 +57,128 @@ func HandleConnections(
 		// In either case, the writer is closed.
 		m.register(conn)
 		go func() {
-			h(conn, w, idGen, connectedClients)
+			defer conn.Close()
+			h(reader(conn), w, idGen, connectedClients)
 			w.Stop() // make sure it's flushed
 			m.remove(conn)
 		}()
 	}
 }
 
+// ErrIllegalTopicFormat is returned if a record did not start with a proper topic prefix.
+var ErrIllegalTopicFormat = errors.New("illegal topic formatting")
+
+// RecordReader emits records that are prefixed with a topic.
+// It returns io.EOF if the underlying record source has no more data.
+type RecordReader func() (record []byte, err error)
+
+// NewDynamicRecordReader returns a record reader that expects each input record from r
+// to start with a space-delimited topic identifier.
+func NewDynamicRecordReader(r io.Reader) RecordReader {
+	br := bufio.NewReader(r)
+
+	return func() ([]byte, error) {
+		for {
+			l, err := br.ReadBytes('\n')
+			if err != nil {
+				return nil, err
+			}
+			// Validate that a correct topic prefix exists.
+			if i := bytes.IndexByte(l, ' '); i < 0 || !isValidTopic(l[:i]) {
+				return nil, ErrIllegalTopicFormat
+			}
+			return l, nil
+		}
+	}
+}
+
+// NewStaticRecordReader returns a record reader that prefixes all input records
+// with a static topic.
+func NewStaticRecordReader(topic string, r io.Reader) (RecordReader, error) {
+	tb := []byte(topic)
+	if !isValidTopic(tb) {
+		return nil, ErrIllegalTopicFormat
+	}
+	tb = append(tb, ' ')
+	br := bufio.NewReader(r)
+
+	return func() ([]byte, error) {
+		for {
+			l, err := br.ReadBytes('\n')
+			if err != nil {
+				return nil, err
+			}
+			return append(tb, l...), nil
+		}
+	}, nil
+}
+
+// isValidTopic ensures that the a topic only contains allowed characters. It implements
+// the regex [a-zA-Z0-9_-]+
+// It takes a byte slice to avoid memory allocations at the caller.
+func isValidTopic(b []byte) bool {
+	for _, c := range b {
+		if !(c == '_' || c == '-' || ('0' <= c && c <= '9') || ('a' <= c && c <= 'z') || ('A' <= c && c <= 'Z')) {
+			return false
+		}
+	}
+	return len(b) > 0
+}
```

Comment:
Hmm, what do you think about being a bit stricter, e.g. `[a-zA-Z][a-zA-Z0-9_-]+`, to reserve `_this` kind of thing for future/system use?

Reply:
Yes, that'd make sense.